### PR TITLE
Set and check a new "user_id" cookie for gameplay

### LINKED ISSUE #242: Isolate online play entries with subdomains
The comp allows for arbitrary Javascript code to be uploaded and run from the main comp domain. This is a substantial security risk. The play online pages should be put on a separate subdomain, and ideally isolated from each other on different subdomains.

### PR DESCRIPTION
**Pull request by jmacdotorg**
This work sets a new cookie named "user_id" on login. It contains, blowfish-encrypted and base64-encoded, the current user's database ID.

When someone wishes to play a game outside of the judging season, the app decrypts this cookie (if present) and loads the user with the resulting ID. If that user is allowed to see this game -- that is, they are the game's author, or have "curator" rights to the comp -- then the game will load for them.

This differs from the app's current behavior, which merely checks the ID of the currently logged-in user.

We do this extra work in order to allow games to run on their own subdomains, which purposefully do not have access to the ifcomp.org login-session cookie. We give this new "user_id" cookie access to all ifcomp.org subdomains, so that games running in these walled-off subdomains can perform this check.

This work, plus some additional Apache-level configuration not in version control, fixes #242.

### CODE REVIEW

**Comment by jmacdotorg** [comment]
Coverall is unhappy because the entirety of Plack::Middleware::IFComp is uncovered, and I added a subroutine to it.

This module does need coverage, but that ain't happening today.

File: IFComp/lib/Plack/Middleware/IFComp.pm [hunk 1]
```diff
@@ -75,36 +94,34 @@ sub call {
...
-            my $session_id = $req->cookies->{ifcomp_session};
+            # Check for an encrypted 'user_id' cookie value to see if the
+            # current user is either the author of this entry or an IFComp
+            # curator. If so, then they're allowed to see the entry early.
+            my $req = Plack::Request->new($env);
+            my $key = $self->config->{blowfish_key};
+
+            use Data::Dumper;
+            warn Dumper( $self->config );
+            warn "And so, the key is " . $self->config->{blowfish_key};
```
Comment (mjmusante) [inline IFComp/lib/Plack/Middleware/IFComp.pm:105]: are these lines being left in?

Comment (jmacdotorg) [reply IFComp/lib/Plack/Middleware/IFComp.pm]: Whoops! Nice catch; lines zapped.

**Review by mjmusante — APPROVED** [review]
looks good to me!

**Review by adherzog — CHANGES_REQUESTED** [review]
I thought we'd decided to set the expiration date of the cookie to the date the cookie is no longer needed (when the current comp opens for judging, I suppose.)

Can we set the cookie only when it's necessary (when the comp is in whatever phase that requires it.) There's no need to set it for all of the judges once the comp opens for judging.

I don't love the config name of 'blowfish_key' - it's specific to a particular algorithm that might change later, and doesn't actually indicate what the key is for. Maybe something like 'user_id_cookie_key' instead?

Add the config key to the 'ifcomp_local.conf-dist' file, please.

**Comment by jmacdotorg** [comment]
> I thought we'd decided to set the expiration date of the cookie to the date the cookie is no longer needed (when the current comp opens for judging, I suppose.)
>
> Can we set the cookie only when it's necessary (when the comp is in whatever phase that requires it.) There's no need to set it for all of the judges once the comp opens for judging.

I changed my mind about these, opting to keep things simple and let the new cookie live for exactly as long as the session cookie instead. This way, if someone logs in to judge a game in October, and then enters a game the following summer on the same session, hey, they'll have a functioning user_id cookie without any need to cycle their login.

So I think I'd like to keep these as they are?

> I don't love the config name of 'blowfish_key' - it's specific to a particular algorithm that might change later, and doesn't actually indicate what the key is for. Maybe something like 'user_id_cookie_key' instead?
>
> Add the config key to the 'ifcomp_local.conf-dist' file, please.

Good suggestions; Done and done.

File: IFComp/lib/IFComp/Controller/Auth.pm [hunk 2]
```diff
@@ -25,6 +25,7 @@ The login handler
 =cut
 
 use IFComp::Form::Login;
+use Crypt::Eksblowfish::Blowfish;
```
Comment (adherzog) [inline IFComp/lib/IFComp/Controller/Auth.pm:28]: I think this needs to be moved to Controller::Root

**Review by adherzog — CHANGES_REQUESTED** [review]
One minor change inline, and then please re-tidy - the CI tests are failing.

Otherwise, looks good to me!

**Review by adherzog — APPROVED** [review]

**Pull request merged**